lib/mongoidable/rspec/controller_matchers.rb

```diff
@@ -0,0 +1,151 @@
+require "rspec/expectations"
+
+module Mongoidable
+  module RSpec
+    module ControllerMatchers
+      extend ::RSpec::Matchers::DSL
+
+      class ControllerHelper < SimpleDelegator
+        def initialize(obj, controller, authorize_pairs)
+          super(obj)
+
+          @controller = controller
+          @authorize_pairs = authorize_pairs
+        end
+
+        def error_message
+          "The #{controller.class.name} did not authenticate the action #{action} against the specified object."
+        end
+
+        def process_action
+          setup_controller
+
+          matchers = authorization_matcher
+          controller._run_process_action_callbacks
+
+          loaded_matcher
+
+          matchers
```

Comment:
So, [this false positive](https://github.com/CardTapp/mortgagemapp/pull/6171/files#diff-8f8416fa87cda8c86da94eb6b8a9c4b1R18) just plain bugged me.

I figured out a way to prevent it.

This involves 2 changes.

Here:
```
          setup_controller

          controller._run_process_action_callbacks
          authorization_matcher
```

In `authorization_matcher`:
```
          authorize_pairs.map do |authorize_action, authorized_object|
            matchers = []
            matcher = have_received(:authorize!).with(authorize_action, authorized_object)
            matcher.setup_expectation(controller)

            matchers << matcher

            unless authorized_object.is_a?(Class)
              matcher = have_received(:authorize!).with(authorize_action, authorized_object.class)
              matcher.setup_negative_expectation(controller)

              matchers << matcher
            end

            matchers
          end.flatten
```

What this does:

* Sets up the controller.
* Calls the action and the pre-filters
* Calls 1-2 expectations based on the test:
  * `expect(controller).to have_received(action, object)`
  * `expect(controller).not_to have_received(action, object.class)` (if `object` is not a Class.)

This prevents the false positive situation we ran into:
```
  authorize!(action, object_class)
  expect(self).to have_received(:authorize!).with(action, object_class_instance) # true
  expect(self).not_to have_received(:authorize!).with(action, object_class) # false
```
```